tests/unittests/unittests.cpp

```diff
@@ -4264,6 +4269,137 @@ class ConcurrentQueueTests : public TestClass<ConcurrentQueueTests>
 		
 		return true;
 	}
+
+	bool acquire_and_signal()
+	{
+		typedef ::moodycamel::LightweightSemaphore semaphore;
+
+		// Test resource acquisition from one other thread.
+		{
+			semaphore s;
+			s.signal(); // Single resource available
+						
+			const auto fnTestSingleAcquire = 
+				[](::moodycamel::LightweightSemaphore &s) -> bool
+				{
+					for (std::size_t k = 0; k < 100; ++k)
+					{
+						ASSERT_OR_FAIL(s.wait(100 * 1000)); // Wait 100 ms
+						s.signal();
+					}
+					
+					return true;
+				};
+
+			std::thread t1(fnTestSingleAcquire, std::ref(s));
+			std::thread t2(fnTestSingleAcquire, std::ref(s));
+
+			// Ideally std::thread would have 'try_join_for' like Boost :(
+			t1.join();
+			t2.join();
+			
+			ASSERT_OR_FAIL(1==s.availableApprox());
+		}
+
+		// Test resource acquisition from multiple threads.
+		{
+			const std::size_t ITERATIONS = 100;
+			const std::array<std::size_t, 2> arrayItemsToWait = {2,3};
+			
+			semaphore s;
+
+			s.signal((100 * arrayItemsToWait[0]) + (100 * arrayItemsToWait[1]));
+			
+			const auto fnTestMultiAcquire =
+				[ITERATIONS](const std::size_t  n, ::moodycamel::LightweightSemaphore &s) -> bool
+				{
+					for (std::size_t k = 0; k < ITERATIONS; ++k)
+					{
+						const ssize_t signedN = static_cast<ssize_t>(n);
+						ASSERT_OR_FAIL(signedN==s.waitMany(n, 100 * 1000)); // Wait 100 ms
+					}
+					
+					return true;
+				};
+				
+			std::packaged_task<bool(std::size_t, moodycamel::LightweightSemaphore&)> task1(fnTestMultiAcquire);
+			std::packaged_task<bool(std::size_t, moodycamel::LightweightSemaphore&)> task2(fnTestMultiAcquire);
+			
+			std::future<bool> f1 = task1.get_future();
+			std::future<bool> f2 = task2.get_future();
+
+			std::thread t1(std::move(task1), arrayItemsToWait[0], std::ref(s));
+			std::thread t2(std::move(task2), arrayItemsToWait[1], std::ref(s));
```

Comment:
Ideally there would be more than just two threads hammering on the semaphore. Also, this test is fine in and of itself, but the case where the semaphore dips below zero and bounces back is where all the tricky logic lies, and is not tested at all.